lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx

```diff
@@ -23,6 +34,69 @@ export type DashboardActivityFiltersProps = {
   onClearSelection?: () => void;
 };
 
+function usePaginatedAPIFetch<
+  Prop extends string,
+  ListType extends any[],
+  FetchType extends { [P in Prop]: ListType } & { pagination: Pagination },
+>(
+  prop: Prop,
+  path: string,
+  filters: Record<string, string | string[] | undefined>,
+): [ListType | undefined, FetchResult<FetchType>, () => void] {
+  const [urlFromPagination, setURLFromPagination] = useState<string>();
+  const [currentList, setCurrentList] = useState<ListType>();
+
+  const fetchResult = useAPIFetch<FetchType>(
+    urlFromPagination ?? path,
+    !urlFromPagination ? filters : undefined,
+  );
+
+  const loadNextPage = useCallback(() => {
+    if (!fetchResult.isLoading && fetchResult.data?.pagination.next) {
+      setURLFromPagination(fetchResult.data.pagination.next);
+    }
+  }, [fetchResult.data?.pagination.next, fetchResult.isLoading]);
+
+  // Every time data is loaded, append the result to the list
+  useEffect(() => {
+    setCurrentList(prev => {
+      if (!fetchResult.data) {
+        return prev;
+      }
+
+      return [...(prev ?? []), ...fetchResult.data[prop]] as ListType;
+    });
+  }, [prop, fetchResult.data]);
```

Comment:
This side effect which updates state should potentially be doable in the form of derived state via `useMemo` or the like.

However, it needs to be able to track its own previous state, and I couldn't make it work in the form of a memoized state which merged something with itself, as the first time is not initialized.

I have the feeling there's a way to do that, but this was the only working way I managed to do.

Reply:
> You shouldn't do a state update inside `useMemo`.

I meant that I tried to replace `useState` + `useEffect` with `useMemo`.

> You can invoke `setCurrentList` on every render and just return the previous value if there is nothing to append.

Yeah, this is perhaps what should be done here. Let me try that.